### PR TITLE
libbeat: support explicit dynamic templates

### PR DESCRIPTION
**Pull request by axw**
## What does this PR do?

Add the field `DynamicTemplate bool` to `mapping.Field`, indicating that the field represents an explicitly named dynamic template. Update Elasticsearch template generation to create dynamic templates from these fields when the target Elasticsearch is 7.13.0+, when support for these dynamic templates was added and the requirement of match criteria was dropped.

Such dynamic templates are suitable only for use in dynamic_templates bulk request parameters and in ingest pipelines; they will have no path or type matching criteria.

## Why is it important?

In APM we would like to be able to dynamically map histogram and summary metrics: https://github.com/elastic/apm-server/issues/3195. To do this we will use the new (7.13.0+) feature of Elasticsearch where dynamic_templates can be specified by name at ingest time, to have explicit control over how dynamic mapping takes place. We will need to predefine dynamic templates for each combination of field type and unit that we support in dynamic metrics.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have made corresponding change to the default configuration files~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: libbeat/template/processor.go [hunk 1]
```diff
@@ -442,14 +453,14 @@ func (p *Processor) object(f *mapping.Field) common.MapStr {
...
 
 type dynamicTemplateKey struct {
-	path      string
+	name      string
 	pathMatch string
 	matchType string
 }
 
-func (p *Processor) addDynamicTemplate(path string, pathMatch string, properties common.MapStr, matchType string) {
+func (p *Processor) addDynamicTemplate(name, pathMatch, matchType string, properties common.MapStr) bool {
```
Comment (exekias) [inline libbeat/template/processor.go:461]: It seems you now return a `bool`, but this is never used, right?

File: libbeat/template/processor.go [hunk 2]
```diff
@@ -120,7 +120,15 @@ func (p *Processor) Process(fields mapping.Fields, state *fieldState, output com
 		}
 
 		if len(indexMapping) > 0 {
-			output.Put(mapping.GenerateKey(field.Name), indexMapping)
+			if field.DynamicTemplate {
```
Comment (exekias) [inline libbeat/template/processor.go:123]: Should we check here that `ObjectTypeMappingType` & `ObjectType` were not provided? This should avoid human mistake if perhaps `dynamic_template` is misunderstood by the dev.

**Review by exekias — DISMISSED** [review]
Thank you! 🚀 left a few comments

File: libbeat/template/processor.go (hunk 1 above)
Comment (axw) [reply libbeat/template/processor.go]: Indeed. I meant to revert that; done now. Thanks!

File: libbeat/template/processor.go (hunk 2 above)
Comment (axw) [reply libbeat/template/processor.go:123]: I've added validation in libbeat/mapping to prevent DynamicTemplate + ObjectTypeParams/ObjectType.

**Comment by jsoriano** [comment]
> In APM we would like to be able to dynamically map histogram and summary metrics: [elastic/apm-server#3195](https://github.com/elastic/apm-server/issues/3195).

Do we expect differences on these dynamic maps between beats? or custom mappings of this kind per integration? If not, perhaps we could hardcode and always install the dynamic mapping for histograms and summaries (and other types we could need in the future), so we are sure that all collectors use the same approach to store these types of metrics. And we don't add more options to fields definition ymls.

**Comment by exekias** [comment]
> Do we expect differences on these dynamic maps between beats? or custom mappings of this kind per integration? If not, perhaps we could hardcode and always install the dynamic mapping for histograms and summaries (and other types we could need in the future), so we are sure that all collectors use the same approach to store these types of metrics. And we don't add more options to fields definition ymls.

I have opened this conversation as a follow up of our latests metrics discussion: https://github.com/elastic/elasticsearch/issues/72536. I think it's ok to keep going with these PRs as we work on getting better alignment

**Comment by jsoriano** [comment]
> > Do we expect differences on these dynamic maps between beats? or custom mappings of this kind per integration? If not, perhaps we could hardcode and always install the dynamic mapping for histograms and summaries (and other types we could need in the future), so we are sure that all collectors use the same approach to store these types of metrics. And we don't add more options to fields definition ymls.
> 
> I have opened this conversation as a follow up of our latests metrics discussion: [elastic/elasticsearch#72536](https://github.com/elastic/elasticsearch/issues/72536). I think it's ok to keep going with these PRs as we work on getting better alignment

I was thinking in the option of including the default dynamic mapping for histograms and summaries in the base index template that libbeat generates, the same way as the [`strings_as_keyword` dynamic mapping](https://github.com/elastic/beats/blob/cdcbe36be36b50791ff4aaf7fa69fc457b675333/libbeat/template/template.go#L370) is included. This way it would be consistent for all agents, and we wouldn't need to add more options to fields definitions.

**Comment by axw** [comment]
@jsoriano I think we should standardise, but I'm not sure about forcing it on every beat like that. As @exekias points out in https://github.com/elastic/elasticsearch/issues/72536, we'll end up with many combinations for a completely general solution. So I think I'd rather wait for that conversation to play out before trying to standardise.

You make a good point about not adding yet more options to `fields.yml` though. Another option I considered was to enable beats to add dynamic templates in code, by adding a non-marshalled field to [libbeat/template.TemplateConfig](https://github.com/elastic/beats/blob/e38e0df73e25b1014df166c613254e6de2619aa3/libbeat/template/config.go#L42-L59). Then in code, apm-server would do something like:

```go
templateConfig.DynamicTemplates = map[string]mapping.Field{
    "histogram_counter_nanos": mapping.Field{Type: "histogram", MetricType: "counter", Unit: "nanos"},
}
```

It would not be possible to define these in `fields.yml`. We would need to come up with another solution for packages, such as Elasticsearch installing a default component template that defines the dynamic templates, and which packages could optionally include in data stream templates.

WDYT?

**Comment by jsoriano** [comment]
> You make a good point about not adding yet more options to `fields.yml` though. Another option I considered was to enable beats to add dynamic templates in code, by adding a non-marshalled field to [libbeat/template.TemplateConfig](https://github.com/elastic/beats/blob/e38e0df73e25b1014df166c613254e6de2619aa3/libbeat/template/config.go#L42-L59). Then in code, apm-server would do something like:
> 
> ```go
> templateConfig.DynamicTemplates = map[string]mapping.Field{
>     "histogram_counter_nanos": mapping.Field{Type: "histogram", MetricType: "counter", Unit: "nanos"},
> }
> ```

I like this option, this would be similar to what I was thinking of always adding these templates, but without forcing this to every beat.  If at some point we find a general solution we can implement it the same way for every beat.

> It would not be possible to define these in `fields.yml`. We would need to come up with another solution for packages, such as Elasticsearch installing a default component template that defines the dynamic templates, and which packages could optionally include in data stream templates.

Are we installing now the `strings_as_keyword` dynamic mapping in packages? If we are we could follow the same approach.

If we are going to wait to confirm a general solution maybe the approach of using `fields.yml` for this is better, specially thinking on giving the same solution in beats and in integrations. But if each package and beat can define its own dynamic mappings for these types we have the risk of having too many variants difficult to migrate when we agree on a general solution.

**Comment by axw** [comment]
> Are we installing now the strings_as_keyword dynamic mapping in packages? If we are we could follow the same approach.

Yes we do: https://github.com/elastic/kibana/blob/f736df990c0bf24c78ca02bb3c06d6276b7b706a/x-pack/plugins/fleet/server/services/epm/elasticsearch/template/template.ts#L407-L418

> If we are going to wait to confirm a general solution maybe the approach of using fields.yml for this is better, specially thinking on giving the same solution in beats and in integrations. But if each package and beat can define its own dynamic mappings for these types we have the risk of having too many variants difficult to migrate when we agree on a general solution.

I can go either way. Speaking only for APM, I don't think it'll be a problem for us to change from one dynamic template name to another as we'll be doing the mapping in an ingest pipeline. Because the template and ingest pipeline are installed together we'll be able to change them at the same time.

@exekias @ruflin do you have an opinion on doing this in code vs. `fields.yml` for right now, and how it should be handled in the future?

**Comment by ruflin** [comment]
I did not dig into the full conversation but for the fields.yml support, I think that is preferrable. Will these templates also be specified by packages in the future? If yes, I think fields.yml is our way to go and we need to bring support for it also to Fleet.

**Comment by exekias** [comment]
I think it's ok to move forward with this. We only have a few entry points where dynamic templates will be used for now, so things should be easy to maintain. Let's not wait on the other conversation as it may take some time to align and get the features we need.

Also an eventual update on namings will be transparent for the user, as we will be updating our code / mappings, but not the UX.

@axw something that would be interesting is to follow up with adding this to the [package spec](https://github.com/elastic/package-spec) and Kibana so we can have dynamic templates in packages too. Do you plan to open issues for that?

**Comment by axw** [comment]
> something that would be interesting is to follow up with adding this to the package spec and Kibana so we can have dynamic templates in packages too. Do you plan to open issues for that?

Will do :+1:

**Review by jsoriano — APPROVED** [review]
Ok, let's go on with this approach, thanks!

**Pull request merged**